headers/crypto/1.2/psa/crypto-wpa3-sae.h

```diff
@@ -0,0 +1,313 @@
+ * // receive commit message
+ * psa_pake_input(operation, #PSA_PAKE_STEP_COMMIT, ...);
+ * \endcode
+ *
+ * If the Hash-To-Element variant is used and a list of rejected groups
+ * is available, it must be provided as a salt:
+ *
+ * \code
+ * // input salt
+ * psa_pake_input(operation, #PSA_PAKE_STEP_SALT, ...);
+ * \endcode
+ *
+ * Then call the following functions in any order:
+ * \code
+ * // set send-confirm counter
+ * psa_pake_input(operation, #PSA_PAKE_STEP_SEND_CONFIRM, ...);
+ * // send confirm message
+ * psa_pake_output(operation, #PSA_PAKE_STEP_CONFIRM, ...);
+ * // receive confirm message
+ * psa_pake_input(operation, #PSA_PAKE_STEP_CONFIRM, ...);
+ * // get key id (optional)
+ * psa_pake_output(operation, #PSA_PAKE_STEP_KEYID, ...);
+ * \endcode
+ * 
+ * Remarks:
+ * \c psa_pake_input(#PSA_PAKE_STEP_SEND_CONFIRM) must be called before
+ * \c psa_pake_output(#PSA_PAKE_STEP_CONFIRM) to set the send-confirm counter.
+ * The #PSA_PAKE_STEP_SEND_CONFIRM and #PSA_PAKE_STEP_CONFIRM steps may be used
+ * multiple times to handle repeated confirm messages with varying counts.
+ *
+ * Finally get the shared secret: 
+ *
+ * \code
+ * // get secret
+ * psa_pake_get_shared_key();
+ * \endcode
+ *
+ * The shared secret produced by WPA3-SAE is pseudorandom.
+ * It can be used directly as an encryption key or as input to a key derivation
+ * operation.
+ */
+#define PSA_ALG_WPA3_SAE_FIXED_BASE             ((psa_algorithm_t) 0x0a000800)
+#define PSA_ALG_WPA3_SAE_FIXED(hash_alg) (PSA_ALG_WPA3_SAE_FIXED_BASE | ((hash_alg) & PSA_ALG_HASH_MASK))
+#define PSA_ALG_WPA3_SAE_GDH_BASE               ((psa_algorithm_t) 0x0a000900)
+#define PSA_ALG_WPA3_SAE_GDH(hash_alg) (PSA_ALG_WPA3_SAE_GDH_BASE | ((hash_alg) & PSA_ALG_HASH_MASK))
+#define PSA_ALG_IS_WPA3_SAE(alg) (((alg) & ~0x000001ff) == PSA_ALG_WPA3_SAE_FIXED_BASE)
+#define PSA_ALG_IS_WPA3_SAE_FIXED(alg) (((alg) & ~PSA_ALG_HASH_MASK) == PSA_ALG_WPA3_SAE_FIXED_BASE)
+#define PSA_ALG_IS_WPA3_SAE_GDH(alg) (((alg) & ~PSA_ALG_HASH_MASK) == PSA_ALG_WPA3_SAE_GDH_BASE)
+
+/** The key confirmation value.
+ * 
+ * This value is used during the key confirmation phase of a PAKE protocol.
+ * The format of the value depends on the algorithm and cipher suite:
+ *
+ * For SPAKE2+ algorithms, the format for both input and output at this step is
+ * the same as the output of the MAC algorithm specified in the cipher suite.
+ *
+ * For PSA_ALG_SRP_6, the format for both input and output at this step is
+ * the same as the output of the hash algorithm specified.
+ *
+ * For WPA3_SAE algorithms, the format for both input and output at this step
+ * is a 2 byte little-endian "send-confirm" counter followed by the output of
+ * the hash algorithm specified.
+ */
+#define PSA_PAKE_STEP_CONFIRM                   ((psa_pake_step_t)0x04)
+
+/** The WPA3-SAE commit step.
+ *
+ * The format for both input and output at this step is a 2 byte number
+ * specifying the group used followed by a scalar and an element of the

```

Comment:
Should we make the _commit-scalar_ and **_COMMIT_ELEMENT_** components of the commit frames separate outputs/inputs (as we do for the separate values in J-PAKE)?

Does concatenating them make it easier for the caller to assemble/disassemble the SAE Commit frames in 802.11?

Reply:
We do not think it makes a big difference for the caller.
However, more STEP constants would be needed.